core/trino-spi/src/main/java/io/trino/spi/type/Decimals.java

```diff
@@ -295,17 +297,44 @@ public static void writeShortDecimal(BlockBuilder blockBuilder, long value)
     public static long rescale(long value, int fromScale, int toScale)
     {
         if (toScale < fromScale) {
-            throw new IllegalArgumentException("target scale must be larger than source scale");
+            // Use MAX_SHORT_PRECISION to fit all precision long values
+            value = shortToShortCast(value,
+                    MAX_SHORT_PRECISION,
+                    fromScale,
+                    MAX_SHORT_PRECISION,
+                    toScale,
+                    longTenToNth(fromScale - toScale),
+                    longTenToNth(fromScale - toScale - 1) * 5);
         }
-        return value * longTenToNth(toScale - fromScale);
+        else {
+            value = value * longTenToNth(toScale - fromScale);
+        }
+
+        if (overflows(value, MAX_SHORT_PRECISION)) {
+            throw new TrinoException(NUMERIC_VALUE_OUT_OF_RANGE,
+                    format("Rescaled value for %s is out of range",
+                            Decimals.toString(value, fromScale)));
+        }
+        return value;
     }
 
     public static BigInteger rescale(BigInteger value, int fromScale, int toScale)
     {
         if (toScale < fromScale) {
-            throw new IllegalArgumentException("target scale must be larger than source scale");
+            value = decodeUnscaledValue(
+                    longToLongCast(
+                            encodeUnscaledValue(value),
+                            MAX_PRECISION,
+                            fromScale,
+                            MAX_PRECISION,
+                            toScale));
         }
-        return value.multiply(bigIntegerTenToNth(toScale - fromScale));
+        else {
+            value = value.multiply(bigIntegerTenToNth(toScale - fromScale));
+        }
+
+        checkOverflow(value);
```

Comment:
Do we need to checkOverflow for both the cases ?

Reply:
I think checking only when we are scaling up should suffice.

Reply:
nit : Can we capture in a separate commit.